xmrstak/backend/cpu/minethd.cpp

```diff
@@ -356,25 +356,35 @@ void minethd::work_main()
 			    either because of network latency, or a socket problem. Since we are
 			    raison d'etre of this software it us sensible to just wait until we have something*/
 
-			while (globalStates::inst().inst().iGlobalJobNo.load(std::memory_order_relaxed) == iJobNo)
+			while (globalStates::inst().iGlobalJobNo.load(std::memory_order_relaxed) == iJobNo)
 				std::this_thread::sleep_for(std::chrono::milliseconds(100));
 
 			consume_work();
 			continue;
 		}
 
+		size_t nonce_ctr = 250;
 		if(oWork.bNiceHash)
-			result.iNonce = calc_nicehash_nonce(*piNonce, oWork.iResumeCnt);
+			result.iNonce = globalStates::inst().calc_start_nonce(*piNonce & 0xFF000000, 4096);
 		else
-			result.iNonce = calc_start_nonce(oWork.iResumeCnt);
+			result.iNonce = globalStates::inst().calc_start_nonce(0, 4096);
 
 		assert(sizeof(job_result::sJobID) == sizeof(pool_job::sJobID));
 		memcpy(result.sJobID, oWork.sJobID, sizeof(job_result::sJobID));
 
-		while(globalStates::inst().inst().iGlobalJobNo.load(std::memory_order_relaxed) == iJobNo)
+		while(globalStates::inst().iGlobalJobNo.load(std::memory_order_relaxed) == iJobNo)
 		{
 			if ((iCount & 0xF) == 0) //Store stats every 16 hashes
 			{
+				if(--nonce_ctr == 0)
```

Comment:
This if condition must be moved out of the statistic counter and `nonce_ctr` must be `4096` at the start.
I tested it and can't measure  any speed impact.
If the condition is within the `iCount` the nonce index space is not fully used.

e.g
- i.  `iCount == 0x6` (or `N * 0x10  + 1`) and `nonce_ctr == 4`
- ii. `iCount` is incremented https://github.com/fireice-uk/xmr-stak/pull/45/files#diff-dcb588b7465f85cba435cdcd930ed1e3R393
- iii. we get a new job or the pool is switched and leave the loop https://github.com/fireice-uk/xmr-stak/pull/45/files#diff-dcb588b7465f85cba435cdcd930ed1e3R375
- iv. `nonce_ctr` set to 256 and we get a new start nonce https://github.com/fireice-uk/xmr-stak/pull/45/files#diff-dcb588b7465f85cba435cdcd930ed1e3R370
- v. now iCount is `0x7`
- vi. incremented `iCount` until `(iCount & 0xF) == 0`
  - if the condition is true `nonce_ctr` is decremented (the first time from 256 to 255 but we ave only calculated 9 shares)

All in all this is a place where many bugs can be coming in. It takes me over 30 minutes to write this short post because, first I thorough that we get duplicated shares because of the asynchronous  decrements/ increment of `iCount` and `nonce_ctr` after a new job use loaded. Than I saw that we only miss some nonces in the index space.
A separate if for `nonce_ctr` will simplify the code and therefore reduce the chance to do something wrong. 

Same for double hash calculation